TestAddIndexFinishImportError is invalid

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
TestAddIndexFinishImportError depends on failpoint `mockFinishImportErr` which is run only when disttask is disabled, after #53233 the case always panics with disttask disabled, the case pass CI because distask is enabled on default now.
```
[2024/05/29 13:20:54.611 +08:00] [ERROR] [misc.go:114] ["panic in the recoverable goroutine"] [label=ddl] [funcInfo=onCreateIndex] [r="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/pkg/util.Recover
    /Users/jujiajia/code/pingcap/tidb/pkg/util/misc.go:118
runtime.gopanic
    /usr/local/go/src/runtime/panic.go:920
runtime.panicmem
    /usr/local/go/src/runtime/panic.go:261
runtime.sigpanic
    /usr/local/go/src/runtime/signal_unix.go:861
github.com/pingcap/tidb/pkg/lightning/backend.(*OpenedEngine).LocalWriter
    /Users/jujiajia/code/pingcap/tidb/pkg/lightning/backend/backend.go:307
github.com/pingcap/tidb/pkg/ddl/ingest.(*engineInfo).newWriterContext
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/ingest/engine.go:224
github.com/pingcap/tidb/pkg/ddl/ingest.(*engineInfo).CreateWriter
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/ingest/engine.go:195
github.com/pingcap/tidb/pkg/ddl.newAddIndexIngestWorker
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/index.go:1680
github.com/pingcap/tidb/pkg/ddl.(*ingestBackfillScheduler).createWorker
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/backfilling_scheduler.go:482
github.com/pingcap/tidb/pkg/ddl.(*ingestBackfillScheduler).setupWorkers.func1
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/backfilling_scheduler.go:400
github.com/pingcap/tidb/pkg/resourcemanager/pool/workerpool.(*WorkerPool[...]).runAWorker
    /Users/jujiajia/code/pingcap/tidb/pkg/resourcemanager/pool/workerpool/workerpool.go:135
github.com/pingcap/tidb/pkg/resourcemanager/pool/workerpool.(*WorkerPool[...]).Start
    /Users/jujiajia/code/pingcap/tidb/pkg/resourcemanager/pool/workerpool/workerpool.go:110
github.com/pingcap/tidb/pkg/ddl.(*ingestBackfillScheduler).setupWorkers
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/backfilling_scheduler.go:403
github.com/pingcap/tidb/pkg/ddl.(*ddlCtx).writePhysicalTableRecord
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/backfilling.go:621
github.com/pingcap/tidb/pkg/ddl.(*worker).addPhysicalTableIndex
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/index.go:1933
github.com/pingcap/tidb/pkg/ddl.(*worker).addTableIndex
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/index.go:1979
github.com/pingcap/tidb/pkg/ddl.runReorgJobAndHandleErr.func1
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/index.go:1041
github.com/pingcap/tidb/pkg/ddl.(*worker).runReorgJob.func1
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/reorg.go:205"] [stack="github.com/pingcap/tidb/pkg/util.Recover
    /Users/jujiajia/code/pingcap/tidb/pkg/util/misc.go:114
runtime.gopanic
    /usr/local/go/src/runtime/panic.go:920
runtime.panicmem
    /usr/local/go/src/runtime/panic.go:261
runtime.sigpanic
    /usr/local/go/src/runtime/signal_unix.go:861
github.com/pingcap/tidb/pkg/lightning/backend.(*OpenedEngine).LocalWriter
    /Users/jujiajia/code/pingcap/tidb/pkg/lightning/backend/backend.go:307
github.com/pingcap/tidb/pkg/ddl/ingest.(*engineInfo).newWriterContext
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/ingest/engine.go:224
github.com/pingcap/tidb/pkg/ddl/ingest.(*engineInfo).CreateWriter
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/ingest/engine.go:195
github.com/pingcap/tidb/pkg/ddl.newAddIndexIngestWorker
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/index.go:1680
github.com/pingcap/tidb/pkg/ddl.(*ingestBackfillScheduler).createWorker
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/backfilling_scheduler.go:482
github.com/pingcap/tidb/pkg/ddl.(*ingestBackfillScheduler).setupWorkers.func1
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/backfilling_scheduler.go:400
github.com/pingcap/tidb/pkg/resourcemanager/pool/workerpool.(*WorkerPool[...]).runAWorker
    /Users/jujiajia/code/pingcap/tidb/pkg/resourcemanager/pool/workerpool/workerpool.go:135
github.com/pingcap/tidb/pkg/resourcemanager/pool/workerpool.(*WorkerPool[...]).Start
    /Users/jujiajia/code/pingcap/tidb/pkg/resourcemanager/pool/workerpool/workerpool.go:110
github.com/pingcap/tidb/pkg/ddl.(*ingestBackfillScheduler).setupWorkers
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/backfilling_scheduler.go:403
github.com/pingcap/tidb/pkg/ddl.(*ddlCtx).writePhysicalTableRecord
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/backfilling.go:621
github.com/pingcap/tidb/pkg/ddl.(*worker).addPhysicalTableIndex
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/index.go:1933
github.com/pingcap/tidb/pkg/ddl.(*worker).addTableIndex
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/index.go:1979
github.com/pingcap/tidb/pkg/ddl.runReorgJobAndHandleErr.func1
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/index.go:1041
github.com/pingcap/tidb/pkg/ddl.(*worker).runReorgJob.func1
    /Users/jujiajia/code/pingcap/tidb/pkg/ddl/reorg.go:205"]

```
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiDB version? (Required)
master
<!-- Paste the output of SELECT tidb_version() -->

--- a/pkg/ddl/backfilling.go
+++ b/pkg/ddl/backfilling.go
@@ -593,7 +593,7 @@ func (dc *ddlCtx) writePhysicalTableRecord(
 	}
 	defer scheduler.close(true)
 	if lit, ok := scheduler.(*ingestBackfillScheduler); ok {
-		if lit.finishedWritingNeedImport() {
+		if lit.importStarted() {
 			return nil
 		}
 	}

--- a/pkg/ddl/backfilling_scheduler.go
+++ b/pkg/ddl/backfilling_scheduler.go
@@ -370,13 +370,13 @@ func newIngestBackfillScheduler(
 	}, nil
 }
 
-func (b *ingestBackfillScheduler) finishedWritingNeedImport() bool {
+func (b *ingestBackfillScheduler) importStarted() bool {
 	job := b.reorgInfo.Job
 	bc, ok := ingest.LitBackCtxMgr.Load(job.ID)
 	if !ok {
 		return false
 	}
-	return bc.FinishedWritingNeedImport()
+	return bc.ImportStarted()
 }
 
 func (b *ingestBackfillScheduler) setupWorkers() error {

--- a/pkg/ddl/ingest/backend.go
+++ b/pkg/ddl/ingest/backend.go
@@ -59,11 +59,11 @@ type BackendCtx interface {
 	// TODO(lance6716): refine the interface to let caller don't need to pass the
 	// indexID, and unify with CollectRemoteDuplicateRows.
 	FinishImport(indexID int64, unique bool, tbl table.Table) error
-	// FinishedWritingNeedImport returns true only when all the engines are finished
-	// writing and only need import. Considering the calling usage of FinishImport,
-	// it will return true after a successful call of FinishImport and may return
-	// true after a failed call of FinishImport.
-	FinishedWritingNeedImport() bool
+	// ImportStarted returns true only when all the engines are finished writing and
+	// import is started by FinishImport. Considering the calling usage of
+	// FinishImport, it will return true after a successful call of FinishImport and
+	// may return true after a failed call of FinishImport.
+	ImportStarted() bool
 
 	CollectRemoteDuplicateRows(indexID int64, tbl table.Table) error
 	FlushController

--- a/pkg/ddl/ingest/engine_mgr.go
+++ b/pkg/ddl/ingest/engine_mgr.go
@@ -111,13 +111,13 @@ func (bc *litBackendCtx) UnregisterEngines() {
 	bc.memRoot.Release(numIdx * (structSizeEngineInfo + engineCacheSize))
 }
 
-// FinishedWritingNeedImport implements BackendCtx.
-func (bc *litBackendCtx) FinishedWritingNeedImport() bool {
+// ImportStarted implements BackendCtx.
+func (bc *litBackendCtx) ImportStarted() bool {
 	if len(bc.engines) == 0 {
 		return false
 	}
 	for _, ei := range bc.engines {
-		if ei.closedEngine != nil {
+		if ei.openedEngine == nil {
 			return true
 		}
 	}

--- a/pkg/ddl/ingest/mock.go
+++ b/pkg/ddl/ingest/mock.go
@@ -119,8 +119,8 @@ func (*MockBackendCtx) UnregisterEngines() {
 	logutil.DDLIngestLogger().Info("mock backend ctx unregister")
 }
 
-// FinishedWritingNeedImport implements BackendCtx interface.
-func (*MockBackendCtx) FinishedWritingNeedImport() bool {
+// ImportStarted implements BackendCtx interface.
+func (*MockBackendCtx) ImportStarted() bool {
 	return false
 }
 

--- a/tests/realtikvtest/addindextest4/ingest_test.go
+++ b/tests/realtikvtest/addindextest4/ingest_test.go
@@ -376,6 +376,7 @@ func TestAddIndexFinishImportError(t *testing.T) {
 	tk.MustExec("create database addindexlit;")
 	tk.MustExec("use addindexlit;")
 	tk.MustExec(`set global tidb_ddl_enable_fast_reorg=on;`)
+	tk.MustExec("set global tidb_enable_dist_task = off;")
 
 	tk.MustExec("create table t (a int primary key, b int);")
 	for i := 0; i < 4; i++ {

